apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,884 @@
+ * @type {Defer}
+ */
+var previousAnimationDefer;
+
+/**
+ * Wait for the end of an animation on `panelElement`.
+ *
+ * @param {Element} panelElement The element to watch.
+ * @returns {Promise} Resolves when the animation ends.
+ */
+function waitForSlideAnimation(panelElement) {
+  if (previousAnimationDefer) {
+    previousAnimationDefer.reject(new Error('A new animation started'));
+  }
 
-    // beforeLeave and beforeEnter can return a rejected promise to prevent
-    // sliding. An exception is never expected though.
-    if (currentPanelObject && currentPanelObject.beforeLeave) {
-      promise = promise.then(
-        currentPanelObject.beforeLeave.bind(currentPanelObject, transitionArgs)
-      );
-    }
+  var defer = Utils.Promise.defer();
+  previousAnimationDefer = defer;
+
+  function onAnimationEnd(e) {
+    panelElement.removeEventListener('animationend', onAnimationEnd);
+    debug('animationend', e && e.type);
+    defer.resolve();
+  }
+
+  panelElement.addEventListener('animationend', onAnimationEnd);
+  setTimeout(onAnimationEnd, 500);
 
-    if (nextPanelObject.beforeEnter) {
-      promise = promise.then(
-        nextPanelObject.beforeEnter.bind(nextPanelObject, transitionArgs)
+  return defer.promise.then(() => previousAnimationDefer = null);
+}
+
+/**
+ * Hides current panel, shows new panel.
+ * It will use an animation if we change panels in the same document.
+ *
+ * @returns {Promise} Resolves when the animation is over, if there is an
+ * animation. Otherwise returns a resolved promise.
+ */
+function switchPanel() {
+  debug('switchPanel()');
+  if (!navigationTransition) {
+    throw new Error(
+      'Trying to switch panel but no transition is started yet.'
+    );
+  }
+
+  var from = navigationTransition.states.from;
+  var to = navigationTransition.states.to;
+
+  var oldView;
+  var oldPanelElement;
+
+  if (from) {
+    oldView = from.panel;
+    oldPanelElement = document.querySelector(`.panel-${oldView.behavior}`);
+    if (!oldPanelElement) {
+      throw new Error(
+        'Couldn\'t find the container element for view ' + oldView.name
       );
     }
+  }
+
+  var newView = to.panel;
+  if (!newView) {
+    throw new Error(
+      'Trying to switch panel, but no new view is defined!'
+    );
+  }
+
+  var newPanelElement = document.querySelector(`.panel-${newView.behavior}`);
+  if (!newPanelElement) {
+    return Promise.reject(new Error(
+      'Couldn\'t find the container element for view ' + newView.name
+    ));
+  }
+
+  var isGoingBack = oldView && oldView.previous === newView.name;
+
+  newPanelElement.style = '';
+
+  var doSlideAnimation =
+    from && from.panel &&
+    from.panel.behavior !== newView.behavior;
 
-    promise = promise.then(function resetPanel() {
-      // While we're sliding, we aren't in any panel
-      currentPanel = null;
+  var animationPromise;
+  if (doSlideAnimation) {
+
+    Object.assign(newPanelElement.style, {
```

Comment:
__nit:__ maybe just assign property directly to the ```style``` object since we have only one property.